### PR TITLE
Store list of apps which cannot be disabled in shipped.json

### PR DESCRIPTION
**Pull request by DeepDiver1975**
@karlitschek @PVince81 @MorrisJobke @icewind1991 @schiesbn please review - THX

### CODE REVIEW

**Comment by karlitschek** [comment]
nice. :+1: I really hope we will only have this two exceptions. But somehow I knew that you will come up with a generic solution ;-) ;-)

**Comment by DeepDiver1975** [comment]
> I really hope we will only have this two exceptions.

there have been discussions in the past to have the provisioning api enabled all the time ..... but this is not really necessary

**Comment by karlitschek** [comment]
provisioning is no longer an issue. But theoretically there would be no reason to remove the option to disable it. We have to always remember that users use ownCloud in interesting ways. No need to remove the option for the admin.

File: lib/private/app/appmanager.php [hunk 1]
```diff
@@ -134,6 +136,9 @@ public function isEnabledForUser($appId, $user = null) {
 	 * @return bool
 	 */
 	private function checkAppForUser($enabled, $user) {
```
Comment (MorrisJobke) [inline lib/private/app/appmanager.php]: This looks :fish:y ... Variable is named `$enabled`but holds the appid?

File: lib/private/app/appmanager.php [hunk 2]
```diff
@@ -134,6 +136,9 @@ public function isEnabledForUser($appId, $user = null) {
...
+		if ($this->isAlwaysEnabled($enabled)) {
+			return true;
+		}
 		if ($enabled === 'yes') {
```
Comment (MorrisJobke) [inline lib/private/app/appmanager.php]: An app named `yes`?

File: lib/private/app/appmanager.php (hunk 1 above)
Comment (MorrisJobke) [inline lib/private/app/appmanager.php]: This method is called two times. One time it is called for each element of `getInstalledAppsValues()`and the other time it is called for the element at index `$appId` of the array of `getInstalledAppsValues()`... looks weird.

Comment (MorrisJobke) [inline lib/private/app/appmanager.php]: Just stepped through it. It's always `yes` except for apps that are only enabled for a group, then it's `["app"]`as string.

**Comment by schiesbn** [comment]
> I really hope we will only have this two exceptions.

Well if we want to reduce the size of core and move more stuff to apps, e.g. in the long run stuff like settings, the app store, etc, it is not unlikely that the list will grow. But I think this wouldn't be bad in general.

**Comment by DeepDiver1975** [comment]
> in the long run stuff like settings, the app store, etc, it is not unlikely that the list will grow

Full ACK on this

**Comment by karlitschek** [comment]
Just wondering when have we decided to move stuff like settings or appstore into an app?

**Comment by DeepDiver1975** [comment]
> Just wondering when have we decided to move stuff like settings or appstore into an app?

Nobody decided anything no point in time - we are exchanging ideas and discuss pros and cons.

**Comment by karlitschek** [comment]
sure. So where are the pros and cons discussed?

**Comment by DeepDiver1975** [comment]
@MorrisJobke please finally review and test - THX

**Comment by DeepDiver1975** [comment]
@icewind1991 @schiesbn @LukasReschke @rullzer please review and test - THX

File: lib/private/app/appmanager.php (hunk 1 above)
Comment (schiesbn) [inline lib/private/app/appmanager.php]: yes, this looks really strange but not really related to this PR, I guess. Do you want to open a separate issue to look into this methods and to clean them up?

**Comment by schiesbn** [comment]
tested, works as expected and code looks good :+1:

File: lib/private/app/appmanager.php [hunk 3]
```diff
@@ -279,4 +281,40 @@ public function getIncompatibleApps($version) {
...
+	}
+
+	private function loadShippedJson() {
+		if (is_null($this->shippedApps)) {
+			$shippedJson = \OC::$SERVERROOT . '/core/shipped.json';
+			if (file_exists($shippedJson)) {
+				$content = json_decode(file_get_contents($shippedJson), true);
+				$this->shippedApps = $content['shippedApps'];
+				$this->alwaysEnabled = $content['alwaysEnabled'];
+			} else {
```
Comment (LukasReschke) [inline lib/private/app/appmanager.php]: What is the reason for this fallback? - Is there a valid case where this makes more sense than the shipped.json file?

Comment (DeepDiver1975) [inline lib/private/app/appmanager.php]: Well it was there from the start.
What else should we do in case the file is not available?

Comment (LukasReschke) [inline lib/private/app/appmanager.php]: IMHO fail hard. No need to have another code path from my PoV. Also the code check to be introduced with 9.0 will require this file to be available for a pass.

Comment (MorrisJobke) [inline lib/private/app/appmanager.php]: > IMHO fail hard. No need to have another code path from my PoV.

Yep. Correct.

**Comment by LukasReschke** [comment]
:+1:

**Pull request merged**